lightningd/nodestatsrpc.c

```diff
@@ -0,0 +1,73 @@
+#include <ccan/autodata/autodata.h>
+#include <common/json.h>
+#include <lightningd/json.h>
+#include <lightningd/jsonrpc.h>
+#include <lightningd/lightningd.h>
+#include <wallet/wallet.h>
+
+static void
+json_add_nodestats(struct json_result *result, char const *field,
+		   const struct nodestats_detail *detail)
+{
+	json_object_start(result, field);
+	json_add_u64(result, "index", detail->index);
+	json_add_pubkey(result, "nodeid", &detail->nodeid);
+	json_add_u64(result, "time_first_seen", detail->time_first_seen);
+	json_add_u64(result, "time_last_seen", detail->time_last_seen);
+	json_add_num(result, "forwarding_failures", detail->forwarding_failures);
+	json_add_num(result, "connect_failures", detail->connect_failures);
+	json_add_num(result, "channel_failures", detail->channel_failures);
+	json_object_end(result);
+}
+
+static void
+json_listnodestats(struct command *cmd,
+		   const char *buffer,
+		   const jsmntok_t *params)
+{
+	struct wallet *wallet = cmd->ld->wallet;
+	jsmntok_t *idtok;
+	struct pubkey id;
+	struct json_result *result;
+	struct nodestats_detail detail;
+	u64 it;
+
+	if (!json_get_params(cmd, buffer, params,
+			     "?id", &idtok,
+			     NULL)) {
+		return;
+	}
+
+	if (idtok && !json_tok_pubkey(buffer, idtok, &id)) {
+		command_fail(cmd, "id is not a public key: '%.*s'",
+			     idtok->end - idtok->start,
+			     buffer + idtok->start);
+		return;
+	}
+
+	result = new_json_result(cmd);
+
+	json_object_start(result, NULL);
+	json_array_start(result, "nodestats");
+	if (idtok) {
+		if (wallet_nodestats_get_by_pubkey(wallet, &detail, &id))
```

Comment:
May return empty despite the caller asking for a specific ID, could return an error instead.

Reply:
ok